### PR TITLE
Link to Reference from autocomplete and parameter help.

### PR DESCRIPTION
**Pull request by microbit-robert**

### CODE REVIEW

**Comment by microbit-robert** [comment]
To do:

- Style links/buttons in autocomplete/parameter help.
- Review the query for map and attempt to return `moduleItems` only when they have reference links.
- Add new e2e tests for linking to Reference content from autocomplete/parameter help.

File: src/editor/codemirror/language-server/documentation.ts [hunk 1]
```diff
@@ -100,10 +101,24 @@ export const subsetMarkdown = (
...
 };
 
+const createStyledAnchorElement = (): HTMLAnchorElement => {
+  const anchor = document.createElement("a");
+  anchor.href = "";
+  anchor.style.fontSize = "var(--chakra-fontSizes-sm)";
+  anchor.style.color = "var(--chakra-colors-brand-600)";
+  anchor.style.display = "block";
+  anchor.style.margin = "0";
+  anchor.style.marginRight = "-0.5rem";
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/documentation.ts:111]: Please check this still makes sense.

File: src/editor/codemirror/language-server/documentation.ts [hunk 2]
```diff
@@ -112,25 +127,63 @@ export const wrapWithDocumentationButton = (
...
   return docsAndActions;
 };
+
+export const getLinkToReference = (
+  id: string,
+  apiReferenceMap: ApiReferenceMap
+): string | undefined => {
+  const idSegments = id.split(".");
+  const pythonModuleName = idSegments[0];
+  const apiId = idSegments.slice(1).join(".");
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/documentation.ts:179]: Please can you consider having some common utility for Python API name manipulation?

File: src/editor/codemirror/language-server/documentation.ts [hunk 3]
```diff
@@ -112,25 +127,63 @@ export const wrapWithDocumentationButton = (
...
+  if (!pythonModuleName && !apiId) {
+    return;
+  }
+  let referenceLink = apiReferenceMap[pythonModuleName]?.[apiId]?.referenceLink;
+  const alternative = apiReferenceMap[pythonModuleName]?.[apiId]?.alternative;
+  if (referenceLink && alternative) {
+    referenceLink += `/${alternative}`;
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/documentation.ts:186]: We should use encodeURIComponent on alternative.

Comment (microbit-robert) [reply src/editor/codemirror/language-server/documentation.ts:186]: Should it also be used on `referenceLink` or is it not required at all. Both are slug fields?

File: src/documentation/mapping/content.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    return undefined;
+  }
+  const map: ApiReferenceMap = {};
+  for (const module of mappingData) {
+    for (const moduleItem of module.pythonModuleItem) {
+      map[module.pythonModuleName] = {
+        ...map[module.pythonModuleName],
+        [moduleItem.pythonApiEntry]: {
+          referenceLink: moduleItem.referenceLink?.slug.current || "",
+          alternative: moduleItem.pythonAlternativeContentLink || "",
```
[the remark this answers is not in the document]
Comment (microbit-robert) [reply src/documentation/mapping/content.ts]: Omit entries where there is no reference link, [`749a683`](https://github.com/microbit-foundation/python-editor-next/pull/806/commits/749a683b9bd459c3166c8e2bc421fdafdc075c16).

File: src/documentation/reference/ReferenceTopicEntry.tsx [hunk 5]
```diff
@@ -43,30 +44,42 @@ const ReferenceTopicEntry = ({
...
+      ));
 
   const hasMore =
     hasCode &&
     (detailContent ||
       contentHasSomeNonCode(content) ||
       (alternatives &&
         contentHasSomeNonCode(
-          alternatives[alternativeIndex as number].content
+          alternatives.find((a) => a.slug.current === alternativeSlug)?.content
```
[the remark this answers is not in the document]
Comment (microbit-robert) [reply src/documentation/reference/ReferenceTopicEntry.tsx]: [`6d05fb1`](https://github.com/microbit-foundation/python-editor-next/pull/806/commits/6d05fb1a863ebc0d298a48d9b9e13cf0ef6f867d)

File: src/editor/codemirror/language-server/documentation.ts (hunk 2 above)
Comment (microbit-robert) [reply src/editor/codemirror/language-server/documentation.ts]: [`7f2a572`](https://github.com/microbit-foundation/python-editor-next/pull/806/commits/7f2a5729e789dadbb060dcb74fbfc5f60b4ad681).

File: src/editor/codemirror/language-server/documentation.ts (hunk 3 above)
Comment (microbit-matt-hillsdon) [reply src/editor/codemirror/language-server/documentation.ts:186]: Yes. Much easier than reasoning about whether it's safe.

File: src/editor/codemirror/language-server/documentation.ts (hunk 1 above)
Comment (microbit-robert) [reply src/editor/codemirror/language-server/documentation.ts]: I can see the reasoning, but I think it's not necessary. Feels like the click target for the link is bigger without the negative margin as it's further in from the edge. [`5cb6a04`](https://github.com/microbit-foundation/python-editor-next/pull/806/commits/5cb6a040978f827858446176643dabd07a9897bf).

File: src/editor/codemirror/CodeMirror.tsx [hunk 6]
```diff
@@ -122,11 +124,20 @@ const CodeMirror = ({
...
-                  },
-                })
+              ? languageServer(
+                  client,
+                  uri,
+                  intl,
+                  logging,
+                  apiReferenceMap.status === "ok"
+                    ? apiReferenceMap.content
+                    : {},
```
Comment (microbit-robert) [inline src/editor/codemirror/CodeMirror.tsx:134]: If the data fails to load, the `Help` button is never displayed. There are no errors, but nothing is logged.

Comment (microbit-robert) [reply src/editor/codemirror/CodeMirror.tsx:134]: @microbit-matt-hillsdon Correction, logging is already handled inside `useContent` in `documentation-hooks.tsx` which we make use of here.

**Pull request merged**